src/compiler/sexp.rs

```diff
@@ -674,68 +674,54 @@ fn parse_sexp_step(loc: Srcloc, p: &SExpParseState, this_char: u8) -> SExpParseR
     }
 }
 
-fn parse_sexp_inner(
+fn parse_sexp_inner<I>(
     start_: Srcloc,
     p_: SExpParseState,
-    n_: usize,
-    s: &[u8],
-) -> Result<Vec<Rc<SExp>>, (Srcloc, String)> {
+    s: I,
```

Comment:
I would expect it to be slower to take an iterator here, instead of a slice. If iterating over the slice is more efficient/simpler in some places, you can always get its iterator. I don't see the benefit of changing the function signature. If anything, knowing that the bytes are contiguous can be used in your advantage in here.

Reply:
The main benefit was that in the language server branch I made iterators over more complex ways of holding bytes that don't need to be turned into a slice.  The individual lines are mutated in the language server so its design reflects that and lets parse_sexp use an interator over a range of line,col to line,col.

Reply:
the other thing to note is that since this is templated, it'll be unrolled for iterator on slice anyway.